### ISSUE TITLE
SelectTagHelper generates invalid markup when using self-closing tag

### ISSUE THREAD

**Issue by zhaparoff** [aspnet/Mvc#4058]
For example:
`<select asp-for="AccountGroupId" asp-items="ViewBag.AvailableAccountGroups" class ="form-control" />`

Generates
`<select class="form-control" data-val="true" data-val-required="The Account Group field is required." id="AccountGroupId" name="AccountGroupId"></select>`

While
`<select asp-for="AccountGroupId" asp-items="ViewBag.AvailableAccountGroups" class ="form-control" ></select>`

Generates correctly into
`<select class="form-control" data-val="true" data-val-required="The Account Group field is required." id="AccountGroupId" name="AccountGroupId">`
`<option value="1">Cash</option>`
`<option selected="selected" value="1003">Debts</option>`
`<option value="1004">Bank Accounts</option>`
`</select>`

**Comment by Muchiachio** [comment]
Self closing select tag is incorrect html, I believe MVC doesn't give us any protection in this regard.

**Comment by zhaparoff** [comment]
That's true.
But I spent a hour today in the morning because of this nonobvious error, wondering why my dropdown has empty list.
Possibly this is the VS issue that cshtml editor does not validate this correctly (Plain HTML editor does)

**Comment by dougbu** [comment]
The MVC tag helpers intentionally do not override what the user wrote in their `.cshtml` file. If the original HTML was invalid (modulo required attributes the tag helper is designed to provide), the generated HTML will be as well.

**Comment by dougbu** [comment]
> Possibly this is the VS issue that cshtml editor does not validate this correctly (Plain HTML editor does)

@zhaparoff is it possible the `.cshtml` editor highlights the incorrect HTML but more subtly than the HTML editor? If not, please file an issue (with screen shots) in the [Tooling](https://github.com/aspnet/Tooling) repo.

**Comment by zhaparoff** [comment]
@dougbu looks like `<select />`, `<div />`, etc. considered as valid HTML in both editors (At least in VS 2015).
Although HTML standard is quite straightforward about this (4.3. Elements):
>The following is a complete list of the **void** elements in HTML: **area, base, br, col, command, embed, hr, img, input, keygen, link, meta, param, source, track, wbr**
...
A **non-void** element **must** have an end tag

**Comment by dougbu** [comment]
@zhaparoff please file an issue (with screen shots [or that spec snippet]) in the [Tooling](https://github.com/aspnet/Tooling) repo.

**Comment by dougbu** [comment]
Couldn't find an existing Tooling issue; filed aspnet/Tooling#398.

**Comment by Eilon** [comment]
@DamianEdwards any thoughts on the behavior of this? We chatted about this a long time ago regarding re-writing a self-closing tag into a close/open tag and vice versa.

**Comment by DamianEdwards** [comment]
Seems strange that in this case we actually expand the tag, but don't fill in the body. If we're truly saying we'll preserve the source, then I'd expect it output as a void tag (albeit invalid) to preserve what the author wrote.

As soon as we start making assumptions at the Tag Helper level as to whether to expand or not, people can't reasonably predict what the output will be.

**Comment by dougbu** [comment]
> in this case we actually expand the tag

I can't reproduce this part of the original report. Hoping that was an unintended edit -- the generated HTML also contains `<select ... />`. @zhaparoff ?

**Comment by zhaparoff** [comment]
@dougbu,
I have made a tests again - seems like Developer Tools in Chrome played a bad joke with me, adding a closing tag.

Output page HTML source:
`<select class="form-control" data-val="true" data-val-required="The Account Group field is required." id="AccountGroupId" name="AccountGroupId" />`

Dev Tools output:
`<select class="form-control" data-val="true" data-val-required="The Account Group field is required." id="AccountGroupId" name="AccountGroupId"></select>`

So tag was modified not by helper, but by the browser tool.
Sorry if I've made mountains out of molehills.
